### PR TITLE
client: do not release the global snaprealm until unmounting

### PR DESCRIPTION
**Pull request by lxbsz**
The global snaprealm would be created and then destroyed immediately
every time when updating it.

Fixes: https://tracker.ceph.com/issues/54362
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by kotreshhr — APPROVED** [review]
LGTM

**Review by ajarr — COMMENTED** [review]
does kclient do something similar?

**Comment by lxbsz** [comment]
> does kclient do something similar?

Yeah and I have already fixed it in kclient.

**Comment by ajarr** [comment]
> > does kclient do something similar?
> 
> Yeah and I have already fixed it in kclient.

Thanks, Xiubo. Got it!

**Review by ajarr — APPROVED** [review]

File: src/client/Client.cc [hunk 1]
```diff
@@ -4814,10 +4814,19 @@ void Client::invalidate_snaprealm_and_children(SnapRealm *realm)
...
+  int count = 1;
+
+  if (!realm) {
     snap_realms[r] = realm = new SnapRealm(r);
-  ldout(cct, 20) << __func__ << " " << r << " " << realm << " " << realm->nref << " -> " << (realm->nref + 1) << dendl;
-  realm->nref++;
+
+    // Do not release the global snaprealm until unmounting.
+    if (r == CEPH_INO_GLOBAL_SNAPREALM) {
+      count++;
```
Comment (vshankar) [inline src/client/Client.cc:4824]: If I am understanding this correctly, you could do away with the local `count` variable, isn't it?

Comment (lxbsz) [reply src/client/Client.cc:4824]: Hi Venky,

Sorry, I don't get this. Currently the `count` is already a local variable.

Actually I can just do the `realm->nref++` here. And then at the end do this again just like what it was before. The reason I add a new `count` variable here is only for the `ldout(ctt, 20)...` below.

Comment (vshankar) [reply src/client/Client.cc:4824]: Well, you can log realm->count before and after too (for debugging I guess) and avoid using `count`.

Comment (lxbsz) [reply src/client/Client.cc:4824]: Do you mean something like:

```
ldout(cct, 20) << __func__ << " " << r << " " << realm << " nref was " << realm->nref << dendl;
...
ldout(cct, 20) << __func__ << " " << r << " " << realm << " nref now is " << realm->nref << dendl;
```

??

Comment (vshankar) [reply src/client/Client.cc:4824]: Right. My point is, you can avoid introducing local `count` if its just for debug purpose.

I know its just a smal local var, but.....

Comment (lxbsz) [reply src/client/Client.cc:4824]: Sure, will update it.

**Review by vshankar — APPROVED** [review]
LGTM

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by lxbsz** [comment]
Just rebased it to the latest code, nothing changed.

**Comment by lxbsz** [comment]
jenkins test make check

**Comment by vshankar** [comment]
jenkins test api

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Apr-14

**Pull request merged**